### PR TITLE
Build with GitHub actions

### PR DESCRIPTION
**Pull request by giorgiosironi**
For https://github.com/libero/publisher/issues/340

Equivalent of the Travis CI one. Would remove the Travis CI build as a follow-up rather than atomically switch.

### CODE REVIEW

**Comment by giorgiosironi** [comment]
Double check appears on https://github.com/libero/article-store/pull/74/commits/f268afbc299956a76a1e9a1f18dd604a17b0324a because the PR was reopened.

**Comment by giorgiosironi** [comment]
Current serial version clocks in at ~3m11s (`11:35:34` to `11:38:45`).
https://github.com/giorgiosironi/article-store/pull/2/checks takes even more time, ~3m35s (`11:29:01` to `11:32:36`).

File: Makefile [hunk 1]
```diff
@@ -34,7 +34,7 @@ start: ## Start the containers
 stop: ## Stop the containers
 	${DOCKER_COMPOSE} down
 
-wait-healthy: ## Wait for the containers to be healthy
+wait-healthy: gitmodules ## Wait for the containers to be healthy
```
Comment (thewilkybarkid) [inline Makefile]: Isn't related?

File: Makefile [hunk 2]
```diff
@@ -52,7 +52,7 @@ exec: ## Run a command on the app container
 		echo "No command provided"; \
 		exit 1; \
 	fi; \
-	${DOCKER_COMPOSE} exec app $(command)
+	${DOCKER_COMPOSE} exec -T app $(command)
```
Comment (thewilkybarkid) [inline Makefile]: What happens when there is a TTY? Should be extracted to a separate PR too.

File: Makefile (hunk 1 above)
Comment (giorgiosironi) [reply Makefile]: Travis CI doesn't require this because it checks out Git submodules on its own, whereas the `actions/checkout` Github Action doesn't so needs this as a dependency

File: Makefile (hunk 2 above)
Comment (giorgiosironi) [reply Makefile]: Strange that removing didn't affect the shell Make target. Will experiment.

File: Makefile (hunk 1 above)
Comment (thewilkybarkid) [reply Makefile]: Oh hang on, script's in the submodule.

File: Makefile (hunk 2 above)
Comment (giorgiosironi) [reply Makefile]: The prompt and niceties like completion disappear when running `sh` with `-T` (which means remove the TTY), so will want to keep the command normal for interactive use outside of CI.

File: Makefile (hunk 1 above)
Comment (giorgiosironi) [reply Makefile]: Extract or leave?

File: Makefile (hunk 2 above)
Comment (giorgiosironi) [reply Makefile]: In https://github.com/libero/article-store/pull/77

File: Makefile (hunk 1 above)
Comment (thewilkybarkid) [reply Makefile]: Extract

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+name: CI
+
+on: [push, pull_request]
```
Comment (thewilkybarkid) [inline .github/workflows/ci.yml]: `push` has the same problem as https://github.com/libero/article-store/pull/61?

File: Makefile (hunk 1 above)
Comment (thewilkybarkid) [reply Makefile]: In fact, looks like it can do the same as Travis? https://github.com/actions/checkout/blob/50fbc622fc4ef5163becd7fab6573eac35f8462e/action.yml#L13-L14

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  build:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v1
+
+    - name: Build Docker images
+      run: |
+        set -e
```
Comment (thewilkybarkid) [inline .github/workflows/ci.yml]: Not needed: https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        TARGET=prod make build
+        TARGET=dev make build
+
+    - name: Lint
+      run: make lint
+      if: always()
```
Comment (thewilkybarkid) [inline .github/workflows/ci.yml]: Meant? Looks like it will be forced to run even if the workflow is cancelled.

**Comment by thewilkybarkid** [comment]
> Current serial version clocks in at ~3m11s (`11:35:34` to `11:38:45`).
> https://github.com/giorgiosironi/article-store/pull/2/checks takes even more time, ~3m35s (`11:29:01` to `11:32:36`).

Probably out of scope for this PR, but would be interested to know what happens when there's a `build` job, then all the others run in parallel (using something like https://stackoverflow.com/a/57877438).

**Comment by thewilkybarkid** [comment]
Should add the badge to the README.

**Comment by giorgiosironi** [comment]
> would be interested to know what happens when there's a build job, then all the others run in parallel

Isn't that what https://github.com/giorgiosironi/article-store/pull/2/checks did? One `build` plus 5 parallel jobs `lint`, `test`, `smoke-tests-prod`, `smoke-tests-dev`, `analyse`. Timestamps available with the hidden button `Show timestamp`.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (giorgiosironi) [reply .github/workflows/ci.yml]: Only alternative I see to execute in spite of other steps failing would be `success() || failure()` which seems confusing?

**Comment by thewilkybarkid** [comment]
> Isn't that what https://github.com/giorgiosironi/article-store/pull/2/checks did? One `build` plus 5 parallel jobs `lint`, `test`, `smoke-tests-prod`, `smoke-tests-dev`, `analyse`. Timestamps available with the hidden button `Show timestamp`.

Ah ok. Looks like that took ~3 minutes 30ish? Not that much slower, and the separate status checks are useful.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (giorgiosironi) [reply .github/workflows/ci.yml]: Yes. https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#pull-request-event-pull_request states:
> Note: By default, a workflow only runs when a pull_request's activity type is `opened`, `synchronize`, or `reopened`. To trigger workflows for more activity types, use the types keyword.

This doesn't seem to cover new commits on an open pull request (nor do the other activity types listed there?). Will try.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+name: CI
+
+on: [pull_request]
```
Comment (giorgiosironi) [inline .github/workflows/ci.yml]: https://github.com/libero/article-store/pull/74/checks?check_run_id=331381010 triggered so I'd leave it like this until it breaks.

Comment (thewilkybarkid) [reply .github/workflows/ci.yml]: Looks like there's more granular options for creation etc: https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onevent_nametypes

**Comment by giorgiosironi** [comment]
seems to be eagerly green despite this not having run on `master`.

**Comment by giorgiosironi** [comment]
https://github.com/libero/article-store/pull/74/commits/3f33297405c533005badd767458ce02cf3f2edfb stopped building on PRs.

**Comment by giorgiosironi** [comment]
> Looks like that took ~3 minutes 30ish? Not that much slower, and the separate status checks are useful.

If the project grows that could become more performant as it scales to larger amount of tests, and we don't pay for resources; at the moment it's 3x the amount of code however.

**Comment by thewilkybarkid** [comment]
Wonder if it supports some of the fancier YAML syntax? 🤔

**Comment by giorgiosironi** [comment]
Tried https://github.com/libero/article-store/pull/79 for parallelization but YAML anchors are not supported.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,43 @@
+name: CI
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
```
Comment (giorgiosironi) [inline .github/workflows/ci.yml:5]: still looking for some documentation about what `synchronize` is. It seems not to be just a merge of `master` into the PR's branch but also adding any commit. Will try removing to see what happens.

**Comment by giorgiosironi** [comment]
No more improvements in the queue, ready for a new pass.

**Comment by thewilkybarkid** [comment]
Going to borrow this PR to try something, will revert if it doesn't work.

**Comment by thewilkybarkid** [comment]
Close to working, but lost the `dev` smoke test.

Feels a bit of a hack though. Tempted to say my preference is to have the separate jobs and live with the duplication. It's a bit annoying, but isn't _that_ bad, and gives us the most useful output. (Separate status checks with sensible names.)

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        test:
+          - make lint
+          - make test
+          - .travis/smoke-test.sh
+          - .travis/analyse.sh
+        include:
```
Comment (giorgiosironi) [inline .github/workflows/ci.yml:62]: reserved keyword in the middle of arbitrary arguments :face_with_head_bandage:

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+          - env: prod
```
Comment (giorgiosironi) [inline .github/workflows/ci.yml]: smoke tests for `prod` have priority over `dev` ones

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+      - name: Download Docker image artifact
+        uses: actions/download-artifact@v1
+        with:
+          name: docker-image-${{ matrix.env || 'dev' }}
+
+      - name: Load Docker image
+        run: docker load -i docker-image-${{ matrix.env || 'dev' }}/article-store.tar
+
+      - name: Run test
+        run: ${{ matrix.test }}
```
Comment (giorgiosironi) [inline .github/workflows/ci.yml:82]: a good thing about a standardized test job is that it prevents logic creeping into this not-actually-YAML file and pushes into scripts and Makefile targets, that can also be executed locally

File: .github/workflows/ci.yml (hunk 9 above)
Comment (giorgiosironi) [reply .github/workflows/ci.yml]: now runs both smoke tests

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    strategy:
+      matrix:
+        test:
+          - make lint
+          - make test
+          - TARGET=dev .travis/smoke-test.sh
+          - TARGET=prod .travis/smoke-test.sh
+          - .travis/analyse.sh
+        include:
+          - env: prod
```
Comment (giorgiosironi) [inline .github/workflows/ci.yml]: though this is `target` rather than `env`

**Comment by giorgiosironi** [comment]
Current status:

- using a matrix to remove all duplication
- testing targets are independent
- their commit status checks are independent
- overhead of parallelism reduced via compression

We could question the name of the status checks to specify them in [branch protection](https://github.com/libero/article-store/settings/branch_protection_rules/11906254). But given how the infrastructure is managed I'd rather specify they all have to pass rather than listing them all there (assuming they will always appear as in progress or failed).

Build performance can follow reasonable patterns but will always be limited because the infrastructure is standardized. Parallelization here helps in the long run with respect to custom hardware.

**Comment by giorgiosironi** [comment]
First hanging build? https://github.com/libero/article-store/runs/335171718

**Comment by giorgiosironi** [comment]
With 3 jobs (part of the same matrix) blocked on the same step.

**Comment by giorgiosironi** [comment]
`Cancel check suite` does turn off the yellow commit statuses.

**Comment by giorgiosironi** [comment]
All jobs were blocked on
```
version-0.12.2: Pulling from tpluscode/hypermedia-app
```
Can only see this now, as the output was not accessible while they were running.

**Comment by giorgiosironi** [comment]
No problems in pulling locally with:
```
docker-compose --file .docker/docker-compose.yml --file .docker/docker-compose.dev.yml pull console
```
```
$ docker image ls | grep hypermedia-app
tpluscode/hypermedia-app                                  version-0.12.2                             6f4fd57ea35f        11 days ago         125MB
generic-hypermedia-app                                    latest                                     6a02fdf92926        13 days ago         125MB
tpluscode/hypermedia-app                                  version-0.12.1                             d6a905c98767        2 weeks ago         125MB
tpluscode/hypermedia-app                                  alpha                                      16099832ee8d        3 weeks ago         124MB
```

**Comment by giorgiosironi** [comment]
Probably a Docker Hub fluke.

**Review by thewilkybarkid — APPROVED** [review]
